### ISSUE TITLE
Question about the audio prompt

### ISSUE THREAD

**Issue by frank613** [e-c-k-e-r/vall-e#8]
Hi Ecker,

I have two questions about the functionality of text and audio prompt in the model "ar+nar-len-llama-8"：

1.  In your implementation, It seems that it is not necessary for the text prompt to be appended with what is spoken in the audio prompt. This is different from what is written the original paper. Is there any special consideration behind this?   Since the code differentiates the prompt and response by using different embeddings, I guess this is the reason that the model doesn't take additional text prompt for the audio prompt.

2. Is the model always trained with 3s audio prompt? Do you think it will harm the quality if I have shorter prompt without repeating it to 3s long?

Thanks,
Xinwei

**Comment by e-c-k-e-r** [comment]
> This is different from what is written the original paper.

I would need to re-read the paper again, but I do remember there being a note about `VALL-E Continuous` where I think that's the modality you're thinking of, where it's something like `<transcription of reference><reference audio><input text><output audio>`.

> Is there any special consideration behind this? 

The [original implementation](https://github.com/enhuiz/vall-e) I used as a base didn't have it that way (along with the proms/resps split), and it proved to not be that necessary. TorToiSe also didn't require it at all, so it seemed normal to me not to need it. I believe this was before Bark came around that did rely on transcribing the reference audio to work.

The other side of it might also be personal bias with not wanting additional cruft with the implementation, as it'd necessitate a transcriber, and puts a restriction on requiring a full audio clip for training (rather than just getting clips).

> Since the code differentiates the prompt and response by using different embeddings, I guess this is the reason that the model doesn't take additional text prompt for the audio prompt.

In a way. My theory is that it coerces the model to reconstruct the speaker from the `prom` by not having an explicit mapping from transcribed text and the reference audio.
* Although I'd argue there's already some sort of relationship where the model *could* source from the input text per the attention mask. I do recall one state of the model where, if there's shared phonemes in the input text with the reference audio, it will pull from it.
* I also do remember past evaluations also having the model can work entirely from the `resps` itself, also implying the model could perfectly well source from itself if needed.

> Is the model always trained with 3s audio prompt?

I believe the model was mostly trained with `cfg.dataset.duration_range = [3.0, 9.0]`, where it'll train with input prompts randomly picked between three and nine seconds, while the more recent-ish post-training sessions picked `[3.0, 6.0]`. I vaguely recall there being some weird trait where it worked better without explicitly training on longer prompts, but I'm not too sure. If I had to point to a specific reason, I think whenever I added in segmented position IDs (where each segment has its own range of position IDs) might have allowed for that to emerge. Maybe.

> Do you think it will harm the quality if I have shorter prompt 

I think two seconds is servicable. Naturally, shorter prompts do reduce speaker similarity, as it does shrink how many utterances can be sourced directly and instead have to be reconstructed (to my understanding). But from what I recall two seconds should be fine.

> without repeating it to 3s long?

If by that you mean with the `trim/repeat`, some checkpoints of the model did benefit where even repeating to fill in a longer duration did help for whatever reason, but if I do remember right it doesn't have any benefits for the more recent weights.

**Comment by frank613** [comment]
Thank you again for the detailed clarification.

Regards,
Xinwei
